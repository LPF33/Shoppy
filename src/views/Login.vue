<template>
  <div v-if="error" class="error">{{ errorMessage }}</div>
  <form @submit.prevent="submit">
    <h1>Login</h1>
    <input
      type="email"
      required
      placeholder="email"
      v-model="email"
      ref="input"
    />
    <input type="password" required placeholder="password" v-model="password" />
    <button type="submit">Submit</button>
  </form>
</template>

<script>
export default {
  name: "Login",
  data() {
    return {
      email: "",
      password: "",
    };
  },
  methods: {
    submit() {
      this.$store.dispatch("signIn", {
        email: this.email,
        password: this.password,
      });
    },
  },
  computed: {
    error() {
      return this.$store.state.error;
    },
    errorMessage() {
      return this.$store.state.errorMessage;
    },
  },
  mounted() {
    this.$refs.input.focus();
  },
};
</script>

<style scoped>
h1 {
  padding-bottom: 10vh;
}

form {
  height: 100%;
  max-width: 500px;
  margin: 0 auto;
  display: flex;
  flex-direction: column;
  justify-content: center;
  text-align: center;
}

button[type="submit"] {
  background-color: #39a9cb;
  color: white;
  padding: 12px 20px;
  cursor: pointer;
  border-radius: 30px;
  margin: 20px 10%;
}
button[type="submit"]:hover {
  background-color: #2940d3;
}
input {
  margin: 5px;
  box-shadow: 0 0 15px 4px rgba(0, 0, 0, 0.06);
  padding: 10px;
  border-radius: 30px;
  margin: 20px 10%;
}
</style>
