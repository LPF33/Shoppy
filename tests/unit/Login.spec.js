import { mount } from "@vue/test-utils";
import Login from "@/views/Login";

let $store = {};

beforeEach(() => {
  $store = {
    state: {
      error: false,
      errorMessage: "",
    },
    dispatch: jest.fn(),
  };
});

describe("Testing the Login component", () => {
  it("Login contains a form with h1, 2x input fields, button", () => {
    const wrapper = mount(Login, {
      global: {
        mocks: { $store },
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
    const form = wrapper.find("form");
    expect(form.exists()).toBeTruthy();
    expect(form.find("h1").exists()).toBeTruthy();
    expect(form.find("h1").text()).toBe("Login");
    const allInputs = form.findAll("input");
    expect(allInputs).toHaveLength(2);
    expect(allInputs[0].attributes("type")).toBe("email");
    expect(allInputs[1].attributes("type")).toBe("password");
    expect(form.find("button").text()).toBe("Submit");
    expect(form.find("button").attributes("type")).toBe("submit");
  });

  it("Fill in email & password, fire Submit event on form, the data is passed to the dispatch function", async () => {
    const wrapper = mount(Login, {
      global: {
        mocks: { $store },
      },
    });
    const form = wrapper.find("form");
    const emailInput = wrapper.find("input[type='email']");
    await emailInput.setValue("test@gmail.com");
    expect(wrapper.find("input[type='email']").element.value).toBe(
      "test@gmail.com"
    );
    const passwordInput = wrapper.find("input[type='password']");
    await passwordInput.setValue("123456");
    expect(wrapper.find("input[type='password']").element.value).toBe("123456");

    const button = form.find("button");
    await button.trigger("submit");
    await wrapper.vm.$nextTick();
    expect($store.dispatch).toHaveBeenCalledTimes(1);
    expect($store.dispatch.mock.calls[0][0]).toBe("signIn");
    expect($store.dispatch.mock.calls[0][1]).toEqual({
      email: "test@gmail.com",
      password: "123456",
    });
  });

  it("Error occurs, should show a div-tag with CSS class 'error' and error Message", () => {
    $store = {
      state: {
        error: true,
        errorMessage: "Error",
      },
      dispatch: jest.fn(),
    };
    const wrapper = mount(Login, {
      global: {
        mocks: { $store },
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
    const errorDiv = wrapper.find(".error");
    expect(errorDiv.exists()).toBeTruthy();
    expect(errorDiv.text()).toBe("Error");
  });

  it("If there is no error, no div-tag with CSS class 'error' rendered on screen", () => {
    const wrapper = mount(Login, {
      global: {
        mocks: { $store },
      },
    });

    const errorDiv = wrapper.find(".error");
    expect(errorDiv.exists()).toBeFalsy();
  });
});
