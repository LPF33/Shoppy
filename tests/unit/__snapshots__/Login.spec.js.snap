// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the Login component Error occurs, should show a div-tag with CSS class 'error' and error Message 1`] = `
<div class="error">Error</div>
<form>
  <h1>Login</h1><input type="email" required="" placeholder="email"><input type="password" required="" placeholder="password"><button type="submit">Submit</button>
</form>
`;

exports[`Testing the Login component Login contains a form with h1, 2x input fields, button 1`] = `
<!--v-if-->
<form>
  <h1>Login</h1><input type="email" required="" placeholder="email"><input type="password" required="" placeholder="password"><button type="submit">Submit</button>
</form>
`;
